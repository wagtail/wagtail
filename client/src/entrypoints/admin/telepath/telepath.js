import Telepath from 'telepath-unpack';

// Prevent double-initialisation if imported twice
// This was introduced because both the sidebar and the page editor
// use telepath, but they use separate hooks for JS dependencies so we
// can't de-duplicate the telepath import.
if (!window.telepath) {
  window.telepath = new Telepath();
}
