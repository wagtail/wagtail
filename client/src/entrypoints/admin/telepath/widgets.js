/* global $ */
import { gettext } from '../../../utils/gettext';

class BoundWidget {
  constructor(element, name, idForLabel, initialState, parentCapabilities) {
    var selector = ':input[name="' + name + '"]';
    this.input = element.find(selector).addBack(selector); // find, including element itself
    this.idForLabel = idForLabel;
    this.setState(initialState);
    this.parentCapabilities = parentCapabilities || new Map();
  }

  getValue() {
    return this.input.val();
  }

  getState() {
    return this.input.val();
  }

  setState(state) {
    this.input.val(state);
  }

  getTextLabel(opts) {
    const val = this.getValue();
    if (typeof val !== 'string') return null;
    const maxLength = opts && opts.maxLength;
    if (maxLength && val.length > maxLength) {
      return val.substring(0, maxLength - 1) + '…';
    }
    return val;
  }

  focus() {
    this.input.focus();
  }

  setCapabilityOptions(capability, options) {
    Object.assign(this.parentCapabilities.get(capability), options);
  }
}

class Widget {
  constructor(html, idPattern) {
    this.html = html;
    this.idPattern = idPattern;
  }

  boundWidgetClass = BoundWidget;

  render(placeholder, name, id, initialState, parentCapabilities) {
    var html = this.html.replace(/__NAME__/g, name).replace(/__ID__/g, id);
    var idForLabel = this.idPattern.replace(/__ID__/g, id);
    var dom = $(html);
    $(placeholder).replaceWith(dom);
    // eslint-disable-next-line new-cap
    return new this.boundWidgetClass(
      dom,
      name,
      idForLabel,
      initialState,
      parentCapabilities,
    );
  }
}
window.telepath.register('wagtail.widgets.Widget', Widget);

class BoundCheckboxInput extends BoundWidget {
  getValue() {
    return this.input.is(':checked');
  }

  getState() {
    return this.input.is(':checked');
  }

  setState(state) {
    // if false, set attribute value to null to remove it
    this.input.attr('checked', state || null);
  }
}

class CheckboxInput extends Widget {
  boundWidgetClass = BoundCheckboxInput;
}
window.telepath.register('wagtail.widgets.CheckboxInput', CheckboxInput);

class BoundRadioSelect {
  constructor(element, name, idForLabel, initialState) {
    this.element = element;
    this.name = name;
    this.idForLabel = idForLabel;
    this.selector = 'input[name="' + name + '"]:checked';
    this.setState(initialState);
  }

  getValue() {
    return this.element.find(this.selector).val();
  }

  getState() {
    return this.element.find(this.selector).val();
  }

  setState(state) {
    this.element.find('input[name="' + this.name + '"]').val([state]);
  }

  focus() {
    this.element.find('input[name="' + this.name + '"]').focus();
  }
}

class RadioSelect extends Widget {
  boundWidgetClass = BoundRadioSelect;
}
window.telepath.register('wagtail.widgets.RadioSelect', RadioSelect);

class BoundSelect extends BoundWidget {
  getTextLabel() {
    return this.input.find(':selected').text();
  }
}

class Select extends Widget {
  boundWidgetClass = BoundSelect;
}
window.telepath.register('wagtail.widgets.Select', Select);

class PageChooserFactory {
  constructor(html, idPattern, config) {
    this.html = html;
    this.idPattern = idPattern;
    this.config = config;
  }

  render(placeholder, name, id, initialState) {
    var html = this.html.replace(/__NAME__/g, name).replace(/__ID__/g, id);
    var dom = $(html);
    $(placeholder).replaceWith(dom);
    /* the PageChooser object also serves as the JS widget representation */
    // eslint-disable-next-line no-undef
    const chooser = new PageChooser(id, null, this.config);
    chooser.setState(initialState);
    return chooser;
  }
}
window.telepath.register('wagtail.widgets.PageChooser', PageChooserFactory);

class AdminAutoHeightTextInput extends Widget {
  render(placeholder, name, id, initialState, parentCapabilities) {
    const boundWidget = super.render(
      placeholder,
      name,
      id,
      initialState,
      parentCapabilities,
    );
    window.autosize($('#' + id));
    return boundWidget;
  }
}
window.telepath.register(
  'wagtail.widgets.AdminAutoHeightTextInput',
  AdminAutoHeightTextInput,
);

class DraftailInsertBlockCommand {
  /* Definition for a command in the Draftail context menu that inserts a block.
   * Constructor args:
   * widget - the bound Draftail widget
   * blockDef - block definition for the block to be inserted
   * addSibling, split - capability descriptors from the containing block's capabilities definition
   */
  constructor(widget, blockDef, addSibling, split) {
    this.widget = widget;
    this.blockDef = blockDef;
    this.addSibling = addSibling;
    this.split = split;

    this.blockMax = addSibling.getBlockMax(blockDef.name);
    this.icon = `#icon-${blockDef.meta.icon}`;
    this.description = blockDef.meta.label;
    this.type = blockDef.name;
  }

  render({ option }) {
    // If the specific block has a limit, render the current number/max alongside the description
    const limitText =
      typeof blockMax === 'number'
        ? ` (${this.addSibling.getBlockCount(this.blockDef.name)}/${
            this.blockMax
          })`
        : '';
    return `${option.description}${limitText}`;
  }

  onSelect({ editorState }) {
    // Reset the current block to unstyled and empty before splitting, so we remove the command prompt if used.
    const result = window.draftail.splitState(
      window.draftail.DraftUtils.resetBlockWithType(editorState, 'unstyled'),
    );
    if (result.stateAfter.getCurrentContent().hasText()) {
      // There is content after the insertion point, so need to split the existing block.
      // Run the split after a timeout to circumvent potential race condition.
      setTimeout(() => {
        if (result) {
          this.split.fn(
            result.stateBefore,
            result.stateAfter,
            result.shouldMoveCommentFn,
          );
        }
        this.addSibling.fn({ type: this.blockDef.name });
      }, 50);
    } else {
      // Set the current block's content to the 'before' state, to remove the '/' separator and
      // reset the editor state (closing the context menu)
      this.widget.setState(result.stateBefore);
      this.addSibling.fn({ type: this.blockDef.name });
    }
  }
}

class DraftailSplitCommand {
  /* Definition for a command in the Draftail context menu that splits the block.
   * Constructor args:
   * widget - the bound Draftail widget
   * split - capability descriptor from the containing block's capabilities definition
   */
  constructor(widget, split) {
    this.widget = widget;
    this.split = split;
    this.description = gettext('Split block');
  }

  icon = '#icon-cut';
  type = 'split';

  onSelect({ editorState }) {
    const result = window.draftail.splitState(
      window.draftail.DraftUtils.resetBlockWithType(editorState, 'unstyled'),
    );
    // Run the split after a timeout to circumvent potential race condition.
    setTimeout(() => {
      if (result) {
        this.split.fn(
          result.stateBefore,
          result.stateAfter,
          result.shouldMoveCommentFn,
        );
      }
    }, 50);
  }
}

class BoundDraftailWidget {
  constructor(input, options, parentCapabilities) {
    this.input = input;
    this.capabilities = new Map(parentCapabilities);
    this.options = options;

    // eslint-disable-next-line no-undef
    const [, setOptions] = draftail.initEditor(
      '#' + this.input.id,
      this.getFullOptions(),
      document.currentScript,
    );
    this.setDraftailOptions = setOptions;
  }

  getValue() {
    return this.input.value;
  }

  getState() {
    return this.input.draftailEditor.getEditorState();
  }

  setState(editorState) {
    this.input.draftailEditor.onChange(editorState);
  }

  getTextLabel(opts) {
    const maxLength = opts && opts.maxLength;
    if (!this.input.value) return '';
    const value = JSON.parse(this.input.value);
    if (!value || !value.blocks) return '';

    let result = '';
    for (const block of value.blocks) {
      if (block.text) {
        result += result ? ' ' + block.text : block.text;
        if (maxLength && result.length > maxLength) {
          return result.substring(0, maxLength - 1) + '…';
        }
      }
    }
    return result;
  }

  focus() {
    setTimeout(() => {
      this.input.draftailEditor.focus();
    }, 50);
  }

  setCapabilityOptions(capability, capabilityOptions) {
    const newCapability = Object.assign(
      this.capabilities.get(capability),
      capabilityOptions,
    );
    this.capabilities.set(capability, newCapability);
    this.setDraftailOptions(this.getFullOptions());
  }

  /**
   * Given a mapping of the capabilities supported by this widget's container,
   * return the options overrides that enable additional widget functionality
   * (e.g. splitting or adding additional blocks).
   * Non-context-dependent Draftail options are available here as this.options.
   */
  getCapabilityOptions(parentCapabilities) {
    const options = {};
    const capabilities = parentCapabilities;
    const split = capabilities.get('split');
    const addSibling = capabilities.get('addSibling');
    let blockCommands = [];
    if (split) {
      const blockGroups =
        addSibling && addSibling.enabled && split.enabled
          ? addSibling.blockGroups
          : [];
      // Create commands for splitting + inserting a block. This requires both the split
      // and addSibling capabilities to be available and enabled
      blockCommands = blockGroups.map(([group, blocks]) => {
        const blockControls = blocks.map(
          (blockDef) =>
            new DraftailInsertBlockCommand(this, blockDef, addSibling, split),
        );
        return {
          label: group || gettext('Blocks'),
          type: `streamfield-${group}`,
          items: blockControls,
        };
      });

      if (split.enabled) {
        blockCommands.push({
          label: 'Actions',
          type: 'custom-actions',
          items: [new DraftailSplitCommand(this, split)],
        });
      }
    }

    options.commands = [
      {
        label: gettext('Rich text'),
        type: 'blockTypes',
      },
      {
        type: 'entityTypes',
      },
      ...blockCommands,
    ];

    return options;
  }

  getFullOptions() {
    return {
      ...this.options,
      ...this.getCapabilityOptions(this.capabilities),
    };
  }
}

class DraftailRichTextArea {
  constructor(options) {
    this.options = options;
  }

  render(container, name, id, initialState, parentCapabilities) {
    const input = document.createElement('input');
    input.type = 'hidden';
    input.id = id;
    input.name = name;

    // If the initialState is an EditorState, rather than serialized rawContentState, it's
    // easier for us to initialize the widget blank and then setState to the correct state
    const initialiseBlank = !!initialState.getCurrentContent;
    input.value = initialiseBlank ? 'null' : initialState;
    container.appendChild(input);

    const boundDraftail = new BoundDraftailWidget(
      input,
      this.options,
      parentCapabilities,
    );

    if (initialiseBlank) {
      boundDraftail.setState(initialState);
    }

    return boundDraftail;
  }
}
window.telepath.register(
  'wagtail.widgets.DraftailRichTextArea',
  DraftailRichTextArea,
);

class BaseDateTimeWidget extends Widget {
  constructor(options) {
    this.options = options;
  }

  render(placeholder, name, id, initialState) {
    const element = document.createElement('input');
    element.type = 'text';
    element.name = name;
    element.id = id;
    placeholder.replaceWith(element);

    this.initChooserFn(id, this.options);

    const widget = {
      getValue() {
        return element.value;
      },
      getState() {
        return element.value;
      },
      setState(state) {
        element.value = state;
      },
      focus(opts) {
        // focusing opens the date picker, so don't do this if it's a 'soft' focus
        if (opts && opts.soft) return;
        element.focus();
      },
      idForLabel: id,
    };
    widget.setState(initialState);
    return widget;
  }
}

class AdminDateInput extends BaseDateTimeWidget {
  initChooserFn = window.initDateChooser;
}
window.telepath.register('wagtail.widgets.AdminDateInput', AdminDateInput);

class AdminTimeInput extends BaseDateTimeWidget {
  initChooserFn = window.initTimeChooser;
}
window.telepath.register('wagtail.widgets.AdminTimeInput', AdminTimeInput);

class AdminDateTimeInput extends BaseDateTimeWidget {
  initChooserFn = window.initDateTimeChooser;
}
window.telepath.register(
  'wagtail.widgets.AdminDateTimeInput',
  AdminDateTimeInput,
);

class ValidationError {
  constructor(messages) {
    this.messages = messages;
  }
}
window.telepath.register('wagtail.errors.ValidationError', ValidationError);
