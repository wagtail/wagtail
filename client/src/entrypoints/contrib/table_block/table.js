/* eslint-disable func-names, dot-notation */

/* global Handsontable */

import $ from 'jquery';
import { hasOwn } from '../../../utils/hasOwn';

function initTable(id, tableOptions) {
  const containerId = id + '-handsontable-container';
  const tableHeaderCheckboxId = id + '-handsontable-header';
  const colHeaderCheckboxId = id + '-handsontable-col-header';
  const tableCaptionId = id + '-handsontable-col-caption';
  const hiddenStreamInput = $('#' + id);
  const tableHeaderCheckbox = $('#' + tableHeaderCheckboxId);
  const colHeaderCheckbox = $('#' + colHeaderCheckboxId);
  const tableCaption = $('#' + tableCaptionId);
  const finalOptions = {};
  let hot = null;
  let dataForForm = null;
  let isInitialized = false;

  const getWidth = function () {
    return $('.w-field--table_input').closest('.w-panel').width();
  };
  const getHeight = function () {
    const tableParent = $('#' + id).parent();
    let htCoreHeight = 0;
    tableParent.find('.htCore').each(function () {
      htCoreHeight += $(this).height();
    });
    return htCoreHeight + tableParent.find('[data-field]').first().height();
  };
  const resizeTargets = ['.handsontable', '.wtHider', '.wtHolder'];
  const resizeHeight = function (height) {
    const currTable = $('#' + id);
    $.each(resizeTargets, function () {
      currTable.closest('[data-field]').find(this).height(height);
    });
  };
  function resizeWidth(width) {
    $.each(resizeTargets, function () {
      $(this).width(width);
    });
    const $field = $('.w-field--table_input');
    $field.width(width);
  }

  try {
    dataForForm = JSON.parse(hiddenStreamInput.val());
  } catch (e) {
    // do nothing
  }

  if (dataForForm !== null) {
    if (hasOwn(dataForForm, 'first_row_is_table_header')) {
      tableHeaderCheckbox.prop(
        'checked',
        dataForForm.first_row_is_table_header,
      );
    }
    if (hasOwn(dataForForm, 'first_col_is_header')) {
      colHeaderCheckbox.prop('checked', dataForForm.first_col_is_header);
    }
    if (hasOwn(dataForForm, 'table_caption')) {
      tableCaption.prop('value', dataForForm.table_caption);
    }
  }

  if (hasOwn(!tableOptions, 'width') || hasOwn(!tableOptions, 'height')) {
    // Size to parent field width if width is not given in tableOptions
    $(window).on('resize', () => {
      hot.updateSettings({
        width: getWidth(),
        height: getHeight(),
      });
      resizeWidth('100%');
    });
  }

  const getCellsClassnames = function () {
    const meta = hot.getCellsMeta();
    const cellsClassnames = [];
    for (let i = 0; i < meta.length; i += 1) {
      if (hasOwn(meta[i], 'className')) {
        cellsClassnames.push({
          row: meta[i].row,
          col: meta[i].col,
          className: meta[i].className,
        });
      }
    }
    return cellsClassnames;
  };

  const persist = function () {
    hiddenStreamInput.val(
      JSON.stringify({
        data: hot.getData(),
        cell: getCellsClassnames(),
        first_row_is_table_header: tableHeaderCheckbox.prop('checked'),
        first_col_is_header: colHeaderCheckbox.prop('checked'),
        table_caption: tableCaption.val(),
      }),
    );
  };

  const cellEvent = function (change, source) {
    if (source === 'loadData') {
      return; // don't save this change
    }

    persist();
  };

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const metaEvent = function (row, column, key, value) {
    if (isInitialized && key === 'className') {
      persist();
    }
  };

  const initEvent = function () {
    isInitialized = true;
  };

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const structureEvent = function (index, amount) {
    resizeHeight(getHeight());
    persist();
  };

  tableHeaderCheckbox.on('change', () => {
    persist();
  });

  colHeaderCheckbox.on('change', () => {
    persist();
  });

  tableCaption.on('change', () => {
    persist();
  });

  const defaultOptions = {
    afterChange: cellEvent,
    afterCreateCol: structureEvent,
    afterCreateRow: structureEvent,
    afterRemoveCol: structureEvent,
    afterRemoveRow: structureEvent,
    afterSetCellMeta: metaEvent,
    afterInit: initEvent,
    // contextMenu set via init, from server defaults
  };

  if (dataForForm !== null) {
    // Overrides default value from tableOptions (if given) with value from database
    if (hasOwn(dataForForm, 'data')) {
      defaultOptions.data = dataForForm.data;
    }
    if (hasOwn(dataForForm, 'cell')) {
      defaultOptions.cell = dataForForm.cell;
    }
  }

  Object.keys(defaultOptions).forEach((key) => {
    finalOptions[key] = defaultOptions[key];
  });
  Object.keys(tableOptions).forEach((key) => {
    finalOptions[key] = tableOptions[key];
  });

  hot = new Handsontable(document.getElementById(containerId), finalOptions);
  window.addEventListener('load', () => {
    // Render the table. Calling render also removes 'null' literals from empty cells.
    hot.render();
    resizeHeight(getHeight());
    resizeWidth(getWidth());
  });
}
window.initTable = initTable;

class TableInput {
  constructor(options, strings) {
    this.options = options;
    this.strings = strings;
  }

  render(placeholder, name, id, initialState) {
    const container = document.createElement('div');
    container.innerHTML = `
      <div className="w-field__wrapper" data-field-wrapper>
        <label class="w-field__label" for="${id}-handsontable-header">${this.strings['Row header']}</label>
        <div class="w-field w-field--boolean_field w-field--checkbox_input" data-field>
          <div id="${id}-handsontable-header-helptext" data-field-help>
            <div class="help">${this.strings['Display the first row as a header.']}</div>
          </div>
          <div className="w-field__input" data-field-input>
            <input type="checkbox" id="${id}-handsontable-header" name="handsontable-header" aria-describedby="${id}-handsontable-header-helptext" />
          </div>
        </div>
      </div>
      <div className="w-field__wrapper" data-field-wrapper>
        <label class="w-field__label" for="${id}-handsontable-col-header">${this.strings['Column header']}</label>
        <div class="w-field w-field--boolean_field w-field--checkbox_input" data-field>
          <div id="${id}-handsontable-col-header-helptext" data-field-help>
            <div class="help">${this.strings['Display the first column as a header.']}</div>
          </div>
          <div className="w-field__input" data-field-input>
            <input type="checkbox" id="${id}-handsontable-col-header" name="handsontable-col-header" aria-describedby="${id}-handsontable-col-header-helptext" />
          </div>
        </div>
      </div>
      <div className="w-field__wrapper" data-field-wrapper>
        <label class="w-field__label" for="${id}-handsontable-col-caption">${this.strings['Table caption']}</label>
        <div class="w-field w-field--char_field w-field--text_input" data-field>
          <div id="${id}-handsontable-col-caption-helptext" data-field-help>
            <div class="help">${this.strings['A heading that identifies the overall topic of the table, and is useful for screen reader users']}</div>
          </div>
          <div className="w-field__input" data-field-input>
            <input type="text" id="${id}-handsontable-col-caption" name="handsontable-col-caption" aria-describedby="${id}-handsontable-col-caption-helptext" />
          </div>
        </div>
      </div>
      <div id="${id}-handsontable-container"></div>
      <input type="hidden" name="${name}" id="${id}" placeholder="${this.strings['Table']}">
    `;
    placeholder.replaceWith(container);

    const input = container.querySelector(`input[name="${name}"]`);
    const options = this.options;

    const widget = {
      getValue() {
        return JSON.parse(input.value);
      },
      getState() {
        return JSON.parse(input.value);
      },
      setState(state) {
        input.value = JSON.stringify(state);
        initTable(id, options);
      },
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      focus() {},
    };
    widget.setState(initialState);
    return widget;
  }
}
window.telepath.register('wagtail.widgets.TableInput', TableInput);
