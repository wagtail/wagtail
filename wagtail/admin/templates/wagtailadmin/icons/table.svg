<svg id="icon-table" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16"><!--! table (solid): Font Awesome Pro 6.4.0 --><path d="M14 1c1.094 0 2 .906 2 2v10c0 1.125-.906 2-2 2H2c-1.125 0-2-.875-2-2V3c0-1.094.875-2 2-2h12ZM7 8V5H2v3h5Zm-5 2v3h5v-3H2Zm7 3h5v-3H9v3Zm5-5V5H9v3h5Z"/></svg>
