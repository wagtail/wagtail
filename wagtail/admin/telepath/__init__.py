from django import forms
from telepath import Adapter, AdapterRegistry, JSContextBase


class WagtailJSContextBase(JSContextBase):
    @property
    def base_media(self):
        # Do not include the telepath.js file in the base media, as it is already included
        # globally in admin_base.html.
        return forms.Media()


class WagtailAdapterRegistry(AdapterRegistry):
    js_context_base_class = WagtailJSContextBase


registry = WagtailAdapterRegistry(telepath_js_path=None)
JSContext = registry.js_context_class


def register(*args, **kwargs):
    return registry.register(*args, **kwargs)


def adapter(js_constructor, base=Adapter):
    """
    Allows a class to implement its adapting logic with a `js_args()` method on the class itself.
    This just helps reduce the amount of code you have to write.

    For example:

        @adapter('wagtail.mywidget')
        class MyWidget():
            ...

            def js_args(self):
                return [
                    self.foo,
                ]

    Is equivalent to:

        class MyWidget():
            ...


        class MyWidgetAdapter(Adapter):
            js_constructor = 'wagtail.mywidget'

            def js_args(self, obj):
                return [
                    self.foo,
                ]
    """

    def _wrapper(cls):
        ClassAdapter = type(
            cls.__name__ + "Adapter",
            (base,),
            {
                "js_constructor": js_constructor,
                "js_args": lambda self, obj: obj.js_args(),
            },
        )

        register(ClassAdapter(), cls)

        return cls

    return _wrapper
