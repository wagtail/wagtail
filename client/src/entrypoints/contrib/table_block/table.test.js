import $ from 'jquery';
window.$ = $;

import '../../admin/telepath/telepath';
import './table';

const TEST_OPTIONS = {
  minSpareRows: 0,
  startRows: 3,
  startCols: 3,
  colHeaders: false,
  rowHeaders: false,
  contextMenu: [
    'row_above',
    'row_below',
    '---------',
    'col_left',
    'col_right',
    '---------',
    'remove_row',
    'remove_col',
    '---------',
    'undo',
    'redo'
  ],
  editor: 'text',
  stretchH: 'all',
  height: 108,
  renderer: 'text',
  autoColumnSize: false,
  language: 'en-us'
};

const TEST_STRINGS = {
  'Row header': 'Row header',
  'Display the first row as a header.': 'Display the first row as a header.',
  'Column header': 'Column header',
  'Display the first column as a header.': 'Display the first column as a header.',
  'Table caption': 'Table caption',
   
  'A heading that identifies the overall topic of the table, and is useful for screen reader users': 'A heading that identifies the overall topic of the table, and is useful for screen reader users',
  'Table': 'Table'
};

const TEST_VALUE = {
  data: [
    ['Test', 'Heading'],
    ['Foo', '123'],
    ['Bar', '456']
  ],
  cell: [],
  first_row_is_table_header: true,
  first_col_is_header: false,
  table_caption: ''
};

// Note: Tests both TableInput and initTable together as this is the only supported way to use them
// It does, however, mock out Handsontable itself
describe('telepath: wagtail.widgets.TableInput', () => {
  // These settings are passed into the table when rendered
  // They are reset for each test so that you can modify them
  let testOptions;
  let testStrings;
  let testValue;
  let handsontableConstructorMock;

  // Call this to render the table block with the current settings
  const render = () => {
    // Create a placeholder to render the widget
    document.body.innerHTML = '<div id="placeholder"></div>';

    // Unpack and render a simple text block widget
    const widgetDef = window.telepath.unpack({
      _type: 'wagtail.widgets.TableInput',
      _args: [testOptions, testStrings]
    });
    return widgetDef.render($('#placeholder'), 'the-name', 'the-id', testValue);
  };

  beforeEach(() => {
    handsontableConstructorMock = jest.fn();
    window.Handsontable = (...args) => { handsontableConstructorMock(...args); };
    window.Handsontable.prototype.render = jest.fn();

    // Reset options, strings, and value for each test
    testOptions = JSON.parse(JSON.stringify(TEST_OPTIONS));
    testStrings = JSON.parse(JSON.stringify(TEST_STRINGS));
    testValue = JSON.parse(JSON.stringify(TEST_VALUE));
  });

  test('it renders correctly', () => {
    render();
    expect(document.body.innerHTML).toMatchSnapshot();
    expect(document.querySelector('input[name="the-name"]').value).toEqual(JSON.stringify(testValue));
  });

  test('Handsontable constructor is called', () => {
    render();
    expect(handsontableConstructorMock.mock.calls.length).toBe(1);
    expect(handsontableConstructorMock.mock.calls[0][0]).toBe(document.getElementById('the-id-handsontable-container'));
    expect(handsontableConstructorMock.mock.calls[0][1].autoColumnSize).toBe(false);
    expect(handsontableConstructorMock.mock.calls[0][1].cell).toEqual([]);
    expect(handsontableConstructorMock.mock.calls[0][1].colHeaders).toBe(false);
    expect(handsontableConstructorMock.mock.calls[0][1].contextMenu).toEqual(testOptions.contextMenu);
    expect(handsontableConstructorMock.mock.calls[0][1].data).toEqual(testValue.data);
    expect(handsontableConstructorMock.mock.calls[0][1].editor).toBe('text');
    expect(handsontableConstructorMock.mock.calls[0][1].height).toBe(108);
    expect(handsontableConstructorMock.mock.calls[0][1].language).toBe('en-us');
    expect(handsontableConstructorMock.mock.calls[0][1].minSpareRows).toBe(0);
    expect(handsontableConstructorMock.mock.calls[0][1].renderer).toBe('text');
    expect(handsontableConstructorMock.mock.calls[0][1].rowHeaders).toBe(false);
    expect(handsontableConstructorMock.mock.calls[0][1].startCols).toBe(3);
    expect(handsontableConstructorMock.mock.calls[0][1].startRows).toBe(3);
    expect(handsontableConstructorMock.mock.calls[0][1].stretchH).toBe('all');
  });

  test('Handsontable.render is called', () => {
    render();
    expect(window.Handsontable.prototype.render.mock.calls.length).toBe(1);
    expect(window.Handsontable.prototype.render.mock.calls[0].length).toBe(0);
  });

  test('translation', () => {
    testStrings = {
      'Row header': 'En-tête de ligne',
      'Display the first row as a header.': 'Affichez la première ligne sous forme d\'en-tête.',
      'Column header': 'En-tête de colonne',
      'Display the first column as a header.': 'Affichez la première colonne sous forme d\'en-tête.',
      'Table caption': 'Légende du tableau',
       
      'A heading that identifies the overall topic of the table, and is useful for screen reader users': 'Un en-tête qui identifie le sujet général du tableau et qui est utile pour les utilisateurs de lecteurs d\'écran',
      'Table': 'Tableau'
    };
    render();
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  test('getValue() returns the current value', () => {
    const boundWidget = render();
    expect(boundWidget.getValue()).toEqual(TEST_VALUE);
  });

  test('getState() returns the current value', () => {
    const boundWidget = render();
    expect(boundWidget.getState()).toEqual(TEST_VALUE);
  });

  test('setState() changes the current state', () => {
    const boundWidget = render();
    testValue.data.push(['Baz', '789']);
    boundWidget.setState(testValue);
    expect(document.querySelector('input[name="the-name"]').value).toEqual(JSON.stringify(testValue));
  });
});
