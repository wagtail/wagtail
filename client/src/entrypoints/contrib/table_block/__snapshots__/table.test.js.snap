// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`telepath: wagtail.widgets.TableInput it renders correctly 1`] = `
"<div>
      <div class=\\"w-field__wrapper\\" data-field-wrapper=\\"\\">
        <label class=\\"w-field__label\\" for=\\"the-id-handsontable-header\\">Row header</label>
        <div class=\\"w-field w-field--boolean_field w-field--checkbox_input\\" data-field=\\"\\">
          <div class=\\"w-field__help\\" id=\\"the-id-handsontable-header-helptext\\" data-field-help=\\"\\">
            <div class=\\"help\\">Display the first row as a header.</div>
          </div>
          <div class=\\"w-field__input\\" data-field-input=\\"\\">
            <input type=\\"checkbox\\" id=\\"the-id-handsontable-header\\" name=\\"handsontable-header\\" aria-describedby=\\"the-id-handsontable-header-helptext\\">
          </div>
        </div>
      </div>
      <div class=\\"w-field__wrapper\\" data-field-wrapper=\\"\\">
        <label class=\\"w-field__label\\" for=\\"the-id-handsontable-col-header\\">Column header</label>
        <div class=\\"w-field w-field--boolean_field w-field--checkbox_input\\" data-field=\\"\\">
          <div class=\\"w-field__help\\" id=\\"the-id-handsontable-col-header-helptext\\" data-field-help=\\"\\">
            <div class=\\"help\\">Display the first column as a header.</div>
          </div>
          <div class=\\"w-field__input\\" data-field-input=\\"\\">
            <input type=\\"checkbox\\" id=\\"the-id-handsontable-col-header\\" name=\\"handsontable-col-header\\" aria-describedby=\\"the-id-handsontable-col-header-helptext\\">
          </div>
        </div>
      </div>
      <div class=\\"w-field__wrapper\\" data-field-wrapper=\\"\\">
        <label class=\\"w-field__label\\" for=\\"the-id-handsontable-col-caption\\">Table caption</label>
        <div class=\\"w-field w-field--char_field w-field--text_input\\" data-field=\\"\\">
          <div class=\\"w-field__help\\" id=\\"the-id-handsontable-col-caption-helptext\\" data-field-help=\\"\\">
            <div class=\\"help\\">A heading that identifies the overall topic of the table, and is useful for screen reader users</div>
          </div>
          <div class=\\"w-field__input\\" data-field-input=\\"\\">
            <input type=\\"text\\" id=\\"the-id-handsontable-col-caption\\" name=\\"handsontable-col-caption\\" aria-describedby=\\"the-id-handsontable-col-caption-helptext\\">
          </div>
        </div>
      </div>
      <div id=\\"the-id-handsontable-container\\"></div>
      <input type=\\"hidden\\" name=\\"the-name\\" id=\\"the-id\\" placeholder=\\"Table\\" value=\\"{&quot;data&quot;:[[&quot;Test&quot;,&quot;Heading&quot;],[&quot;Foo&quot;,&quot;123&quot;],[&quot;Bar&quot;,&quot;456&quot;]],&quot;cell&quot;:[],&quot;first_row_is_table_header&quot;:true,&quot;first_col_is_header&quot;:false,&quot;table_caption&quot;:&quot;&quot;}\\">
    </div>"
`;

exports[`telepath: wagtail.widgets.TableInput translation 1`] = `
"<div>
      <div class=\\"w-field__wrapper\\" data-field-wrapper=\\"\\">
        <label class=\\"w-field__label\\" for=\\"the-id-handsontable-header\\">En-tête de ligne</label>
        <div class=\\"w-field w-field--boolean_field w-field--checkbox_input\\" data-field=\\"\\">
          <div class=\\"w-field__help\\" id=\\"the-id-handsontable-header-helptext\\" data-field-help=\\"\\">
            <div class=\\"help\\">Affichez la première ligne sous forme d'en-tête.</div>
          </div>
          <div class=\\"w-field__input\\" data-field-input=\\"\\">
            <input type=\\"checkbox\\" id=\\"the-id-handsontable-header\\" name=\\"handsontable-header\\" aria-describedby=\\"the-id-handsontable-header-helptext\\">
          </div>
        </div>
      </div>
      <div class=\\"w-field__wrapper\\" data-field-wrapper=\\"\\">
        <label class=\\"w-field__label\\" for=\\"the-id-handsontable-col-header\\">En-tête de colonne</label>
        <div class=\\"w-field w-field--boolean_field w-field--checkbox_input\\" data-field=\\"\\">
          <div class=\\"w-field__help\\" id=\\"the-id-handsontable-col-header-helptext\\" data-field-help=\\"\\">
            <div class=\\"help\\">Affichez la première colonne sous forme d'en-tête.</div>
          </div>
          <div class=\\"w-field__input\\" data-field-input=\\"\\">
            <input type=\\"checkbox\\" id=\\"the-id-handsontable-col-header\\" name=\\"handsontable-col-header\\" aria-describedby=\\"the-id-handsontable-col-header-helptext\\">
          </div>
        </div>
      </div>
      <div class=\\"w-field__wrapper\\" data-field-wrapper=\\"\\">
        <label class=\\"w-field__label\\" for=\\"the-id-handsontable-col-caption\\">Légende du tableau</label>
        <div class=\\"w-field w-field--char_field w-field--text_input\\" data-field=\\"\\">
          <div class=\\"w-field__help\\" id=\\"the-id-handsontable-col-caption-helptext\\" data-field-help=\\"\\">
            <div class=\\"help\\">Un en-tête qui identifie le sujet général du tableau et qui est utile pour les utilisateurs de lecteurs d'écran</div>
          </div>
          <div class=\\"w-field__input\\" data-field-input=\\"\\">
            <input type=\\"text\\" id=\\"the-id-handsontable-col-caption\\" name=\\"handsontable-col-caption\\" aria-describedby=\\"the-id-handsontable-col-caption-helptext\\">
          </div>
        </div>
      </div>
      <div id=\\"the-id-handsontable-container\\"></div>
      <input type=\\"hidden\\" name=\\"the-name\\" id=\\"the-id\\" placeholder=\\"Tableau\\" value=\\"{&quot;data&quot;:[[&quot;Test&quot;,&quot;Heading&quot;],[&quot;Foo&quot;,&quot;123&quot;],[&quot;Bar&quot;,&quot;456&quot;]],&quot;cell&quot;:[],&quot;first_row_is_table_header&quot;:true,&quot;first_col_is_header&quot;:false,&quot;table_caption&quot;:&quot;&quot;}\\">
    </div>"
`;
